packages/core/src/memory/memory.ts

```diff
@@ -0,0 +1,243 @@
+import { Settings } from "../global";
+import type { ChatMessage, LLM } from "../llms";
+import { extractText } from "../utils";
+import type { DefaultAdapters } from "./adapter";
+import { LlamaIndexAdapter, type MessageAdapter } from "./adapter/base";
+import { VercelMessageAdapter } from "./adapter/vercel";
+import { BaseMemory, DEFAULT_TOKEN_LIMIT_RATIO } from "./base";
+import type { MemoryMessage } from "./types";
+
+const DEFAULT_TOKEN_LIMIT = 4096;
+
+export class Memory<
+  TAdapters extends Record<string, MessageAdapter<unknown>> = Record<
+    string,
+    never
+  >,
+> extends BaseMemory {
+  private messages: MemoryMessage[] = [];
+  private tokenLimit: number = DEFAULT_TOKEN_LIMIT;
+  private adapters: TAdapters & DefaultAdapters;
+
+  constructor(
+    messages: MemoryMessage[] = [],
+    tokenLimit?: number,
+    customAdapters?: TAdapters,
+  ) {
+    super();
+    this.messages = messages;
+    this.tokenLimit = tokenLimit ?? DEFAULT_TOKEN_LIMIT;
+
+    this.adapters = {
+      ...customAdapters,
+      llamaindex: LlamaIndexAdapter,
+      vercel: new VercelMessageAdapter(),
+    } as TAdapters & DefaultAdapters;
+  }
+
+  async add(
+    message: unknown,
+    options: { adapter?: keyof TAdapters } = {},
+  ): Promise<void> {
+    let llamaMessage: ChatMessage | null = null;
+
+    // Add with adapter provided
+    if (options.adapter) {
+      const adapter = this.adapters[options.adapter];
+      if (!adapter) {
+        throw new Error(
+          `No adapter registered for type "${options.adapter.toString()}"`,
+        );
+      }
+      if (adapter.isCompatible(message)) {
+        llamaMessage = adapter.toLlamaIndex(message);
+      } else {
+        throw new Error(
+          `The message is not a valid ${options.adapter.toString()} and no compatible adapter was found.`,
+        );
+      }
+    }
+
+    const { llamaindex, ...otherAdapters } = this.adapters;
+
+    // Try to find a compatible adapter among the other adapters
+    for (const key in otherAdapters) {
+      const adapter = otherAdapters[key as keyof typeof otherAdapters];
+      if (adapter?.isCompatible(message)) {
+        llamaMessage = adapter.toLlamaIndex(message);
+        break;
+      }
+    }
+
+    // If no compatible adapter is found, try the llamaindex adapter
+    if (!llamaMessage && llamaindex?.isCompatible(message)) {
+      llamaMessage = llamaindex.toLlamaIndex(message);
+    }
+
+    if (llamaMessage) {
+      this.messages.push(llamaMessage);
+    } else {
+      console.warn(
+        "The message is not a valid ChatMessage and no compatible adapter was found. The message will be ignored.",
+        message,
```

Comment:
better throw an error and:
```suggestion
        "None of the adapters (${apapters.join(",")} can handle the message.",
        message,
```